CMSgov/onemac
# Save table filter and column visibility states

Story: https://qmacbis.atlassian.net/browse/OY2-23135

### Details

This PR updates the app so that it saves filter state and hidden column state when a user changes filter and column show/hide values on both the SPA and WAIVER tables independently. On reload of the page, the saved states are set as the default, so filtration and configuration can be saved in a browser instance for a user.

### Changes

- localStorage (read/write): `onemac-SPA-tableFiltersSavedState`, `onemac-SPA-columnHiddenSavedState`, `onemac-WAIVER-tableFiltersSavedState`, `onemac-WAIVER-columnHiddenSavedState`
- update `react-table.useTable` default table settings to include `hiddenColumns` and `filters` populated by the saved state or empty arrays if no state is saved
- removes `TEMP_onReset` as it was no longer used after implementing this feature

### Implementation Notes

Filters and hidden columns are all handled via `react-table` state inside the application, so none of our storage methods actually touch our filtering; every instance of the app may apply its own unique filters and column configs. All we are doing is extracting the most recently updated states and saving a copy to reload from when the browser comes back to the URL or refreshes the page for any reason.

Hidden Columns are stored as an array of strings, first in the tab's Session, then the same value overwrites the Local store. This is to prevent edits directly being made to a single source of truth from multiple tabs.

Filters are an array of objects, and because they're quite a bit easier to handle thanks to `react-table` managing that array for us, all filter changes do is overwrite the last object in the Local store, no Session needed.

#### More on the sessionStorage to localStorage thing

To ensure multiple tabs would not create a mutated source of truth in `localStorage`, we write any edited data, such as the array containing hidden column values, to `sessionStorage` in the particular tab, then use that edited value to overwrite whatever is in `localStorage`. The resulting UX is as such:

- Change filters in Tab A, switch to Tab B
- Refresh Tab B and see filters from Tab A applied
- Apply filters in Tab B, switch to Tab C
- Refresh Tab C and see filters from Tab B
- Change filters in Tab A, switch to Tab C
- Refresh Tab C and see filters from Tab A

In short, the most recent tab's saved filters and column show/hide states are the global source of truth for a reload in any tab.

### Test Plan

1. Use the UI to set up filters and column configs for the SPA table on your Dashboard
2. Change tabs to Waivers, and back to SPAs -- you should have persisted UI state including filtration and column setup
3. Refresh to ensure they load properly from storage
4. Change tabs to Waivers and repeat Step 1
5. Change tabs to SPAs, and back to Waivers -- you should have persisted UI state again
6. Repeat Step 3
7. Logout and back in
8. Navigate to dashboard and observe a reset of your filters and column setup
